biz.aQute.bndlib/src/aQute/bnd/osgi/ActivelyClosingClassLoader.java

```diff
@@ -212,15 +214,18 @@ public Class<?> loadClass(String name) throws ClassNotFoundException {
 
 				Bundle bundle = FrameworkUtil.getBundle(ActivelyClosingClassLoader.class);
 				if (bundle == null) {
+					// not running inside OSGi...
 					throw nfe;
 				}
-				try {
-					return bundle.loadClass(name);
-				} catch (ClassNotFoundException anotherNfe) {
-					Bundle system = bundle.getBundleContext()
-						.getBundle(0);
-					return system.loadClass(name);
-				}
+				Optional<Class<?>> bundleClass = tryLoadFromBundle(name, bundle, nfe).or(() -> {
+					BundleContext bundleContext = bundle.getBundleContext();
+					if (bundleContext == null) {
+						// not an active bundle, nothing more we can do here...
+						return Optional.empty();
+					}
+					return tryLoadFromBundle(name, bundleContext.getBundle(0), nfe);
+				});
+				return bundleClass.orElseThrow(() -> nfe);
```

Comment:
I think this is a lot more readable:

```
			try {
				return super.loadClass(name);
			} catch (ClassNotFoundException nfe) {
				Bundle bundle = FrameworkUtil.getBundle(ActivelyClosingClassLoader.class);
				BundleContext c;
				if (bundle == null || (c = bundle.getBundleContext()) == null)
					throw nfe;

				try {
					return bundle.loadClass(name);
				} catch (ClassNotFoundException anotherNfe) {
					Bundle system = c.getBundle(0);
					return system.loadClass(name);
				}
			}
```

In general the rule we have about commenting in bnd is not to do it, the code should speak. So if you want to comment something, try making a temporary variable that elucidates the code. It never ceases to amaze me how many of the comments were not valid anymore after 20 years. Also, private methods should use private comments (/* nor /**) or no comments. Again, the the goal is that the code is clear enough.

Although in general I think it is really good to abstract behavior, in this case, the `tryLoadFromBundle` seems to introduce just more complexity and it does not seem useful outside this class. Notice that if you have a BundleContext, you can always get bundle 0 (unless the framework is starting down but that would be the least of your problems then.)